### PR TITLE
core: batch get region size

### LINKED ISSUE #7248: PD processes heartbeat slowly after scaling within a large cluster.
## Bug Report

PD processes heartbeat slowly after scaling within a large cluster.

After scaling, 7.1.x will estimate the progress of scale, the estimate will **hold the lock too long.**

https://github.com/tikv/pd/blob/3f1a6882b2be47767d4a308a305e8da94c90fef4/server/cluster/cluster.go#L1943-L1951

Here `GetRegionSizeByRange` will iterate over many regions.

stack info in 7.1.1:
```
14708261976110 151990 @ 0xf0ada9 0x171c036 0x20d30b1 0x20d2bd6 0x20d171c 0x20c72d4 0xefb901
#	0xf0ada8	sync.(*RWMutex).RUnlock+0x28							/usr/local/go/src/sync/rwmutex.go:119
#	0x171c035	github.com/tikv/pd/pkg/core.(*RegionsInfo).GetRegionSizeByRange+0x175		/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/pd/pkg/core/region.go:1538
#	0x20d30b0	github.com/tikv/pd/server/cluster.(*RaftCluster).calculateRange+0x2b0		/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/pd/server/cluster/cluster.go:1851
#	0x20d2bd5	github.com/tikv/pd/server/cluster.(*RaftCluster).getThreshold+0x3f5		/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/pd/server/cluster/cluster.go:1825
#	0x20d171b	github.com/tikv/pd/server/cluster.(*RaftCluster).checkStores+0x41b		/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/pd/server/cluster/cluster.go:1752
#	0x20c72d3	github.com/tikv/pd/server/cluster.(*RaftCluster).runNodeStateCheckJob+0xd3	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/pd/server/cluster/cluster.go:539
```

### PR DESCRIPTION
**Pull request by rleungx**
### What problem does this PR solve?

Issue Number: Close #7248.

### What is changed and how does it work?

```commit-message
```

### Check List

Tests

- Unit test

### Release note

```release-note
None.
```

### CODE REVIEW

File: pkg/core/region.go [hunk 1]
```diff
@@ -1609,17 +1609,34 @@ func (r *RegionsInfo) ScanRegionWithIterator(startKey []byte, iterator func(regi
 }
 
 // GetRegionSizeByRange scans regions intersecting [start key, end key), returns the total region size of this range.
-func (r *RegionsInfo) GetRegionSizeByRange(startKey, endKey []byte) int64 {
-	r.t.RLock()
-	defer r.t.RUnlock()
+func (r *RegionsInfo) GetRegionSizeByRange(startKey, endKey []byte, limit int) int64 {
```
Comment (lhy1024) [inline pkg/core/region.go:1612]: Do we need to add bench for it?

**Review by lhy1024 — COMMENTED** [review]
fix 7248?

**Comment by nolouch** [comment]
can we do some simulate test?

File: server/cluster/cluster.go [hunk 2]
```diff
@@ -108,6 +108,7 @@ const (
 	// minSnapshotDurationSec is the minimum duration that a store can tolerate.
 	// It should enlarge the limiter if the snapshot's duration is less than this value.
 	minSnapshotDurationSec = 5
+	scanRegionLimit        = 1000
```
Comment (nolouch) [inline server/cluster/cluster.go:111]: How about making it a constant instead of being a parameter in `GetRegionSizeByRange`?

**Comment by nolouch** [comment]
I added a benchmark and ran it in my local MBP, it shows about 1000% improvement.

master
```

go test -benchmem -run=^$ -bench ^BenchmarkRandomSetRegionWithGetRegionSizeByRange$ github.com/tikv/pd/pkg/core --tags=intest

goos: darwin
goarch: arm64
pkg: github.com/tikv/pd/pkg/core
BenchmarkRandomSetRegionWithGetRegionSizeByRange-12    	   88456	     12543 ns/op	       0 B/op	       0 allocs/op


Running tool: /opt/homebrew/bin/go test -benchmem -run=^$ -bench ^BenchmarkRandomSetRegionWithGetRegionSizeByRangeParallel$ github.com/tikv/pd/pkg/core --tags=intest

goos: darwin
goarch: arm64
pkg: github.com/tikv/pd/pkg/core
BenchmarkRandomSetRegionWithGetRegionSizeByRangeParallel-12    	   22857	     58492 ns/op	     755 B/op	      15 allocs/op
PASS
ok  	github.com/tikv/pd/pkg/core	14.837s

```
this branch:

```
Running tool: /opt/homebrew/bin/go test -benchmem -run=^$ -bench ^BenchmarkRandomSetRegionWithGetRegionSizeByRange$ github.com/tikv/pd/pkg/core --tags=intest

goos: darwin
goarch: arm64
pkg: github.com/tikv/pd/pkg/core
BenchmarkRandomSetRegionWithGetRegionSizeByRange-12    	  601136	      1928 ns/op	    4878 B/op	      35 allocs/op
PASS



Running tool: /opt/homebrew/bin/go test -benchmem -run=^$ -bench ^BenchmarkRandomSetRegionWithGetRegionSizeByRangeParallel$ github.com/tikv/pd/pkg/core --tags=intest

goos: darwin
goarch: arm64
pkg: github.com/tikv/pd/pkg/core
BenchmarkRandomSetRegionWithGetRegionSizeByRangeParallel-12    	  241456	      5818 ns/op	   12736 B/op	     103 allocs/op
PASS
ok  	github.com/tikv/pd/pkg/core	16.234s



```

**Review by nolouch — APPROVED** [review]

**Comment by nolouch** [comment]
PTAL @rleungx  @lhy1024

**Pull request merged**